common/Geometry2d/Rect.cpp

```diff
@@ -18,6 +25,125 @@ bool Rect::intersects(const Rect& other) const {
     }
 }
 
+int Rect::CohenSutherlandCode(const Point& other) const{
+    int code;
+    double x,y;
+    x = other.x();
+    y = other.y();
+
+    code = INSIDE;          // initialised as being inside of [[clip window]]
+    if (x < minx())           // to the left of clip window
+        code |= LEFT;
+    else if (x > maxx())      // to the right of clip window
+        code |= RIGHT;
+    if (y < miny())           // below the clip window
+        code |= BOTTOM;
+    else if (y > maxy())      // above the clip window
+        code |= TOP;
+
+    return code;
+}
+
+bool Rect::intersects_(const Segment& other, Point* intr1, Point* intr2) const{
+
+    //Code aggressively borrowed from wikipedia entry Cohen-Sutherland Line Clipping
+    Point p0 = other.pt[0];
+    double x0 = p0.x();
+    double y0 = p0.y();
+    Point p1 = other.pt[1];
+    double x1 = p1.x();
+    double y1 = p1.y();
+
+    int outcode0 = CohenSutherlandCode(p0);
+    int outcode1 = CohenSutherlandCode(p1);
+    
+    //intr1, intr2 = nullptr;
+    Point **nextPoint = &intr1;
```

Comment:
`intr1` can be a nullptr if it's not passed into the function. This would break the program. You should have two function variables if you really want to keep their state then just set intr1 and intr2 to their correct values if they are non-null at the end

Reply:
This wouldn't need to be a double pointer then. You could just make it a list and append it

Reply:
done